frontend/src/app/main/data/workspace/groups.cljs

```diff
@@ -150,6 +153,43 @@
         (pcb/change-parent parent-id children idx-in-parent)
         (pcb/remove-objects [(:id frame)]))))
 
+
+(defn- clone-component-shapes-changes
+  [changes shape objects]
+  (let [[new-shape new-shapes _]
+        (ctst/clone-object shape
+                           (:parent-id shape)
+                           objects)
+
+        new-shape-id (:id new-shape)
+        new-shapes (->> new-shapes
+                        (filter #(not= (:id %) new-shape-id))
+                        (map #(if (= new-shape-id (:parent-id %))
+                                (assoc % :parent-id (:parent-id shape))
+                                %)))]
+    (reduce
+     (fn [changes shape]
+       (pcb/add-object changes shape))
+     changes
+     new-shapes)))
+
+(defn remove-component-changes
+  [it page-id shape objects state]
+  (let [data (get state :workspace-data)
+        page (wsh/lookup-page state)
+        file (wsh/get-local-file state)
```

Comment:
It's not mandatory, but I prefer that helper functions don't receive the whole app state, but only the elements they really need. I think it's better to leave state management only to event functions.

For example, I'll define this function to receive file-data and file. And extract the page from the file with common.types helpers.

Reply:
Changed